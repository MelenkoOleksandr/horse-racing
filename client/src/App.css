@keyframes greeting-letters {
  0% {
    letter-spacing: 1px;
  }

  50% {
    letter-spacing: 5px;
  }

  100% {
    letter-spacing: 1px;
  }
}

.container {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  flex-direction: column;
  padding: 0;
  width: 900px;
}

.greeting {
  width: 80%;
  flex-direction: column;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-bottom: 20px;
}

.user-warning {
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: center;
  width: 100%;
  font-size: 18px;
  margin-bottom: 20px;
}

.greeting-title {
  text-align: center;
  font-size: 36px;
  letter-spacing: 1px;
  margin-bottom: 00px;
  transition: 0.3s;
  animation: greeting-letters;
  animation-duration: 2s;
  animation-iteration-count: 3
}

.start-btn {
  cursor: pointer;
  padding: 10px 30px;
  border: none;
  border-radius: 10px;
  font-size: 20px;
  color: rgb(249, 246, 246);
  background-color: #8cc28c;
}

.start-btn:hover {
  background-color: #97d297;
  box-shadow: 0 0 5px black;
}

.start-btn:disabled {
  background-color: gray;
  cursor: default;
  box-shadow: none;
}

.horses-field {
  width: 100%;
  min-height: 80%;
  border: 10px solid black;
}


.winner {
  font-size: 30px;
  font-weight: bold;
  margin-top: 20px;
}

.winner-name {
  color: red;
}

.user-choice__container {
  width: 100%;
  text-align: center;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.user-choice {
  font-size: 20px;
  margin: 20px 0;
}

.user-choice span {
  color: red;
}

.user-result {
  font-size: 28px;
  font-weight: bold;
}
@media (max-width: 992px) {
  .container {
    width: 550px;
  }
}


@media (max-width: 576px) {
  .container {
    width: 100%;
  }
  .greeting {
    width: 100%;
  }
  .horses-field {
    width: 100%;
    border: none;
  }

  .greeting-title {
    font-size: 24px;
    animation:none;
  }

  .user-warning {
    box-sizing: border-box;
    font-size: 14px;
    font-weight: 400;
    padding: 0 20px;
  }

  .start-btn {
    padding: 10px 25px;
    font-size: 14px;
  }
  .winner {
    font-size: 24px;
  }
  .user-choice {
    margin: 10px 0;
  }
  .user-result {
    font-size: 24px;
  }
}