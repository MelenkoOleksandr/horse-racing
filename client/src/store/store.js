import { configureStore } from "@reduxjs/toolkit";
import raceSlice from './raceSlice/raceSlice';

export default configureStore({
    reducer: raceSlice
})