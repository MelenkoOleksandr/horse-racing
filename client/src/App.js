import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { socket } from './index';
import { setHorses, setRaceStarted, setWinner } from './store/raceSlice/raceSlice';
import Greeting from './components/Greeting/Greeting';
import Horses from './components/Horses';
import Winner from './components/Winner/Winner';
import { allHorsesFinishedRace, findWinnerIndex } from './utils/race';

import './App.css';

export const MAX_DISTANCE = 1000

function App() {
  const dispatch = useDispatch()
  const winner = useSelector(state => state.winner)

  useEffect(() => {
    socket.on('ticker', horses => {
      dispatch(setHorses(horses))

      if (allHorsesFinishedRace(horses)) {
        // If the server will be fixed to multiple races | now app works correctly only on the first time then listeners stacks 
        // dispatch(setRaceStarted(false))
        return
      }

      const winnerIndex = findWinnerIndex(horses)
      if (winner !== -1 || winnerIndex) {
        dispatch(setWinner(winnerIndex))
      }

      return () => {
        socket.off('ticker')
      }
    })
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const startRace = () => {
    socket.emit("start")
    dispatch(setRaceStarted(true))
  }

  return (
    <div className="container">
      <Greeting startRace={startRace} />
      <Horses />
      <Winner />
    </div>
  );
}

export default App;
