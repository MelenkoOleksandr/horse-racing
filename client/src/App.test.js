import App from "./App"
import {fireEvent, render} from "@testing-library/react"
import store from "./store/store"
import { Provider } from "react-redux"

it("App renders", () => {
    const { asFragment } = render(<Provider store={store}><App /></Provider>)
    expect(asFragment()).toMatchSnapshot()
})

describe('Greeting test', () => {
    it('Button is disabled if race have started', () => {
        const greeting = render(<Provider store={store}><App /></Provider>)
        const btn = greeting.getByTestId("greeting-btn")
        fireEvent(btn, new MouseEvent('click'))
        expect(btn).not.toBeDisabled()
    })
})
