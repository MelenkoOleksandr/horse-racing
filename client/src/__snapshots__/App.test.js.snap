// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="greeting"
    >
      <h3
        class="greeting-title"
      >
        Welcome to the horse race!
      </h3>
      <button
        class="start-btn"
        data-testid="greeting-btn"
      >
        Start race
      </button>
    </div>
  </div>
</DocumentFragment>
`;
